### PR TITLE
Reset state on logout

### LINKED ISSUE #243: Ticket state is not cleared on signout
If a user is signed in, the logs out and another user signs in, the tickets of the first user will show up.

This is caused by the ticket cubit not being reset on logout, due to it being registered in our service locator

### PR DESCRIPTION
**Pull request by TTA777**
Closes #243 

Resets the state of all cubits registered in the service locator on logout

### CODE REVIEW

**Review by marfavi — CHANGES_REQUESTED** [review]
The Your tickets section doesn't get updated after a purchase

File: lib/cubits/tickets/tickets_cubit.dart [hunk 1]
```diff
@@ -13,8 +13,12 @@ class TicketsCubit extends Cubit<TicketsState> {
   TicketsCubit(this._ticketRepository) : super(const TicketsLoading());
 
   Future<void> getTickets() async {
+    // Only fetch tickets from backend if they are not loaded already
+    if (state is TicketsLoaded) {
```
Comment (marfavi) [inline lib/cubits/tickets/tickets_cubit.dart:17]: I'm curious, when is this check relevant?

Comment (TTA777) [reply lib/cubits/tickets/tickets_cubit.dart:17]: Without that check, I was seeing that the tickets got refreshed whenever I navigated to another page in the app from the bottom bar. That seemed wasteful, hence the check. You can try removing it locally and seeing for yourself if my description does not make sense.

Comment (marfavi) [reply lib/cubits/tickets/tickets_cubit.dart:17]: Seems like TicketsLoading is emitting when navigating to another tab, which is a bug that should be fixed

Comment (marfavi) [reply lib/cubits/tickets/tickets_cubit.dart:17]: Seems like getTickets (also fetchReceipts) are called on each bottom bar navigation.

Following code in home_page.dart should be refactored:

```dart
        BlocProvider.value(
          value: sl.get<TicketsCubit>()..getTickets(),
        ),
        BlocProvider.value(
          value: sl.get<ReceiptCubit>()..fetchReceipts(),
        ),
```

I can look into it.

Comment (marfavi) [reply lib/cubits/tickets/tickets_cubit.dart:17]: #247 should fix it

File: lib/widgets/components/tickets/tickets_section.dart [hunk 2]
```diff
@@ -14,64 +13,59 @@ class TicketSection extends StatelessWidget {
...
-                //FIXME: display error
-                return const Text('error');
-              }
+                      ),
+                    )
+                    .toList(),
+              );
+            } else if (state is TicketsError) {
+              //FIXME: display error
+              return const Text('error');
```
Comment (jonasanker) [inline lib/widgets/components/tickets/tickets_section.dart:62]: We should show something more descriptive than this :P

File: lib/widgets/components/tickets/tickets_section.dart [hunk 3]
```diff
@@ -14,64 +13,59 @@ class TicketSection extends StatelessWidget {
...
+              ReceiptOverlay.of(context).show(
+                receipt: state.receipt,
+                isTestEnvironment:
+                    envState is EnvironmentLoaded && envState.isTestEnvironment,
+              );
+            }
+          },
+          builder: (context, state) {
+            if (state is TicketsLoading) {
+              return const Text('loading');
```
Comment (jonasanker) [inline lib/widgets/components/tickets/tickets_section.dart:37]: We should refer this string from the String Constants class

File: lib/cubits/authentication/authentication_cubit.dart [hunk 4]
```diff
@@ -41,6 +44,9 @@ class AuthenticationCubit extends Cubit<AuthenticationState> {
 
   Future<void> unauthenticated() async {
     await _storage.clearAuthenticatedUser();
+    //Reset all cubits registered in the service locator. Otherwise their state is persisted to other user logins
+    sl<TicketsCubit>().resetCubit();
+    sl<ReceiptCubit>().resetCubit();
```
Comment (jonasanker) [inline lib/cubits/authentication/authentication_cubit.dart:49]: Is this really the best practice? See https://stackoverflow.com/questions/60124310/how-to-reset-to-the-initial-state-value-whenever-i-close-my-app-using-flutter-bl

**Review by jonasanker — CHANGES_REQUESTED** [review]

File: lib/cubits/authentication/authentication_cubit.dart (hunk 4 above)
Comment (marfavi) [reply lib/cubits/authentication/authentication_cubit.dart:49]: #247 addresses this

File: lib/widgets/components/tickets/tickets_section.dart (hunk 2 above)
Comment (marfavi) [reply lib/widgets/components/tickets/tickets_section.dart]: Check latest commit

File: lib/widgets/components/tickets/tickets_section.dart (hunk 3 above)
Comment (marfavi) [reply lib/widgets/components/tickets/tickets_section.dart]: Replaced with shimmering effect

File: lib/cubits/tickets/tickets_cubit.dart (hunk 1 above)
Comment (marfavi) [reply lib/cubits/tickets/tickets_cubit.dart]: https://github.com/AnalogIO/coffeecard_app/pull/247 merged, fixed

File: lib/cubits/authentication/authentication_cubit.dart (hunk 4 above)
Comment (marfavi) [reply lib/cubits/authentication/authentication_cubit.dart]: https://github.com/AnalogIO/coffeecard_app/pull/247 merged, fixed

**Review by jonasanker — APPROVED** [review]
Lgtm

**Review by marfavi — APPROVED** [review]

**Pull request merged**